### PR TITLE
Refine handling of CanThrow capabilities

### LINKED ISSUE #13816: Safer exceptions with many cases in catch
## Compiler version

3.1.1-RC1-bin-20211020-afaac17-NIGHTLY

## Minimized code

```Scala
import language.experimental.saferExceptions

class Ex1 extends Exception("Ex1")
class Ex2 extends Exception("Ex2")

def foo1(i: Int): Unit throws Ex1 throws Ex2 =
   if i > 0 then throw new Ex1 else throw new Ex2

def foo2(i: Int): Unit throws (Ex1 | Ex2) =
   if i > 0 then throw new Ex1 else throw new Ex2

def foo3(i: Int)(using CanThrow[Ex1], CanThrow[Ex2]) =
   if i > 0 then throw new Ex1 else throw new Ex2

def foo4(i: Int)(using CanThrow[Ex1])(using CanThrow[Ex2]) =
   if i > 0 then throw new Ex1 else throw new Ex2

def foo5(i: Int)(using CanThrow[Ex1 | Ex2]) =
   if i > 0 then throw new Ex1 else throw new Ex2

def test(): Unit =
  try {
    foo1(1)
    foo2(1) // error
    foo3(1)
    foo4(1)
    foo5(1) // error
  } catch {
    case _: Ex1 =>
    case _: Ex2 => 
  }
```

## Output

```scala
[error] -- Error: /Users/mpalka/Projects/dotty-test/tmp/src/main/scala/test.scala:24:11 
[error] 24 |    foo2(1) // error
[error]    |           ^
[error]    |      The capability to throw exception Ex1 | Ex2 is missing.
[error]    |      The capability can be provided by one of the following:
[error]    |       - A using clause `(using CanThrow[Ex1 | Ex2])`
[error]    |       - A `throws` clause in a result type such as `X throws Ex1 | Ex2`
[error]    |       - an enclosing `try` that catches Ex1 | Ex2
[error]    |
[error]    |      The following import might fix the problem:
[error]    |
[error]    |        import unsafeExceptions.canThrowAny
[error]    |
[error] -- Error: /Users/mpalka/Projects/dotty-test/tmp/src/main/scala/test.scala:27:11 
[error] 27 |    foo5(1) // error
[error]    |           ^
[error]    |      The capability to throw exception Ex1 | Ex2 is missing.
[error]    |      The capability can be provided by one of the following:
[error]    |       - A using clause `(using CanThrow[Ex1 | Ex2])`
[error]    |       - A `throws` clause in a result type such as `X throws Ex1 | Ex2`
[error]    |       - an enclosing `try` that catches Ex1 | Ex2
[error]    |
[error]    |      The following import might fix the problem:
[error]    |
[error]    |        import unsafeExceptions.canThrowAny
[error]    |
[error] two errors found
[error] two errors found
```

## Expectation

The compiler should be able to figure out that all the declared exceptions are handled. Interestingly adding `case _: (Ex1 | Ex2) => ` inside catch makes the compiler stop complaining

### LINKED ISSUE #13846: Wrongly synthesized CanThrow for exceptions from Java stdlib
## Compiler version

3.1.2-RC1-bin-20211027-435207d-NIGHTLY

## Minimized code

```Scala
import language.experimental.saferExceptions

def foo(): Int throws ArithmeticException = 1 / 0

def test(): Unit =
  try
    foo()
  catch
    case _: ArithmeticException => println("Caught")
```

## Output

```scala
[error] The capability to throw exception ArithmeticException is missing.
[error] The capability can be provided by one of the following:
[error]  - A using clause `(using CanThrow[ArithmeticException])`
[error]  - A `throws` clause in a result type such as `X throws ArithmeticException`
[error]  - an enclosing `try` that catches ArithmeticException
[error] 
[error] The following import might fix the problem:
[error] 
[error]   import unsafeExceptions.canThrowAny
[error] 
[error]     foo()
[error]          ^
```

## Expectation

This should compile successfully.

When this gets compiled with `-Xprint:typer`, one can see that the synthesized `CanThrow` instance is actually
```
final lazy given erased val evidence$2: CanThrow[Nothing] = ???
```
instead of the expected `CanThrow[ArithmeticException]`.
The same problems occurs e.g. for `IndexOutOfBoundsException`.

### LINKED ISSUE #13849: CanThrow capability shouldn't be synthesized for nonexhaustive catch clauses
## Compiler version

3.1.1-RC1

## Minimized code

```Scala
import annotation.experimental
import language.experimental.saferExceptions

@experimental
case class Ex(i: Int) extends Exception(s"Exception: $i")

@experimental
def foo(): Unit throws Ex = throw Ex(1)

@experimental
object Main:
  def main(args: Array[String]): Unit =
    try
      foo()
    catch
      case _: Ex if false => println("Caught")
```

## Output

```scala
[error] Ex: Exception: 1
[error]         at Ex$.apply(test.scala:4)
[error]         at test$package$.foo(test.scala:8)
[error]         at Main$.main(test.scala:14)
[error]         at Main.main(test.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
```

## Expectation

This should not compile. The compiler should not synthesize given instances of `CanThrow` if it cannot prove at compile time that all subcases for a given exception type are handled.
Beside boolean guards we should also handle extractors like
```scala
    catch
      case Ex(0) => println("Caught")
```

### LINKED ISSUE #13864: Wrongly synthesized CanThrow for exceptions with an invariant type parameter
## Compiler version

3.1.2-RC1-bin-20211101-49b6c08-NIGHTLY

## Minimized code

```Scala
import language.experimental.saferExceptions

case class Ex[A](a: A) extends Exception(s"Ex: $a")

def foo[A](a: A): Unit throws Ex[A] = throw new Ex(a)

def test(): Unit =
  try
    foo(1)
  catch
    case Ex(i: Int) => println("Caught an Int")
```

## Output

```scala
[error] The capability can be provided by one of the following:
[error]  - A using clause `(using CanThrow[Ex[Int]])`
[error]  - A `throws` clause in a result type such as `X throws Ex[Int]`
[error]  - an enclosing `try` that catches Ex[Int]
[error] 
[error] The following import might fix the problem:
[error] 
[error]   import unsafeExceptions.canThrowAny
[error] 
[error]     foo(1)
[error]
```

## Expectation

This should compile successfully.
If one tries to compile this code with `-Xprint:typer` one can see that the entire try/catch block gets desugared to
```scala
      try 
        {
          final lazy given erased val evidence$2: CanThrow[Ex[A$1]] = ???
          foo[Int](1).apply(/* missing */summon[CanThrow[Ex[Int]]])
        }
       catch 
        {
          case Ex.unapply[A$1 @ A$1](i @ _:Int):Ex[A$1] => 
            println("Caught an Int")
        }
```
where `A$1` does not get substituted with `Int` as expected.
Interestingly, the error does not occur if we define `case class Ex[+A](a: A)` with a covariant type parameter

### PR DESCRIPTION
**Pull request by odersky**
1. Disallow throws clauses over RuntimeExceptions
2. Don't generate CanThrow capabilities for catches with guards

Fixes #13864
Fixes #13846
Fixes #13849

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -1936,14 +1939,20 @@ class Typer extends Namer
...
+      if (tycon == defn.andType)
         checkedArgs = checkedArgs.mapconserve(arg =>
           checkSimpleKinded(checkNoWildcard(arg)))
-      else if (tpt1.symbol == defn.orType)
+      else if (tycon == defn.orType)
         checkedArgs = checkedArgs.mapconserve(arg =>
           checkSimpleKinded(checkNoWildcard(arg)))
+      else if tycon == defn.throwsAlias
+          && checkedArgs.length == 2
+          && checkedArgs(1).tpe.derivesFrom(defn.RuntimeExceptionClass)
```
Comment (prolativ) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1951]: Is there really a gain in explicitly preventing users from declaring throwing `RuntimeException`s? And even if we disallow methods like
```scala
def foo(): Int throws ArithmeticException = ???
```
then
```scala
def foo()(using CanThrow[ArithmeticException]): Int = ???
```
will be still valid, right? So this somehow defeats the purpose IMHO

**Comment by prolativ** [comment]
This doesn't fix https://github.com/lampepfl/dotty/issues/13816 (I'll push the fix after this gets merged to avoid conflicts).
I guess you meant https://github.com/lampepfl/dotty/issues/13864 instead

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:1951]: We won't generate a capability for it, so it's better to communicate that early. But I think ding this for throws clauses is enough. I don't want to start imposing restrictions on general implicit parameters.

File: tests/neg/i13846.check [hunk 2]
```diff
@@ -0,0 +1,17 @@
+-- Error: tests/neg/i13846.scala:3:22 ----------------------------------------------------------------------------------
+3 |def foo(): Int throws ArithmeticException = 1 / 0 // error
+  |                      ^^^^^^^^^^^^^^^^^^^
+  |                      throws clause cannot be defined for RuntimeException
+-- Error: tests/neg/i13846.scala:7:9 -----------------------------------------------------------------------------------
+7 |    foo() // error
+  |         ^
+  |         The capability to throw exception ArithmeticException is missing.
+  |         The capability can be provided by one of the following:
+  |          - A using clause `(using CanThrow[ArithmeticException])`
```
Comment (prolativ) [inline tests/neg/i13846.check:10]: If we don't synthesize instances of `CanThrow` for subtypes of `RuntimeException` then this message is rather misleading than helpful, as propagating this capability with using clauses won't make the compilation succeed unless someone creates such an instance by themself. And `X throws ArithmeticException` won't even compile

**Comment by prolativ** [comment]
Then at least maybe we could extend the generic `implicitNotFound` message for `CanThrow` with a note that this won't work if `E` is a subtype of `RuntimeException`? Also this should be explicitly stated in the docs that you cannot declare throwing `RuntimeException`s.
Anyway, if someone (maybe in a library) defines a method with `(using CanThrow[E])` if E <:< RuntimeException (which would still be possible) an then someone else would like to use this method while having saferExceptions enabled, what would be the steps to make this work?

**Review by prolativ — APPROVED** [review]

**Pull request merged**